contracts/l2/token/L2GraphToken.sol

```diff
@@ -0,0 +1,258 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+pragma solidity ^0.7.6;
+
+import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20BurnableUpgradeable.sol";
+import "@openzeppelin/contracts/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+
+import "../../upgrades/GraphUpgradeable.sol";
+import "../../token/GraphToken.sol";
+import "../../arbitrum/IArbToken.sol";
+import "../../governance/Governed.sol";
+
+/**
+ * @title GraphTokenUpgradeable contract
+ * @dev This is the implementation of the ERC20 Graph Token.
+ * The implementation exposes a Permit() function to allow for a spender to send a signed message
+ * and approve funds to a spender following EIP2612 to make integration with other contracts easier.
+ *
+ * The token is initially owned by the deployer address that can mint tokens to create the initial
+ * distribution. For convenience, an initial supply can be passed in the constructor that will be
+ * assigned to the deployer.
+ *
+ * The governor can add contracts allowed to mint indexing rewards.
+ *
+ * Note this is an exact copy of the original GraphToken contract, but using
+ * initializer functions and upgradeable OpenZeppelin contracts instead of
+ * the original's constructor + non-upgradeable approach.
+ */
+contract GraphTokenUpgradeable is GraphUpgradeable, Governed, ERC20Upgradeable, ERC20BurnableUpgradeable {
+    using SafeMath for uint256;
+
+    // -- EIP712 --
+    // https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#definition-of-domainseparator
+
+    bytes32 private constant DOMAIN_TYPE_HASH =
+        keccak256(
+            "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract,bytes32 salt)"
+        );
+    bytes32 private constant DOMAIN_NAME_HASH = keccak256("Graph Token");
+    bytes32 private constant DOMAIN_VERSION_HASH = keccak256("0");
+    bytes32 private constant DOMAIN_SALT =
+        0xe33842a7acd1d5a1d28f25a931703e5605152dc48d64dc4716efdae1f5659591; // Randomly generated salt
```

Comment:
Not sure about this, it seemed safer to use a different salt for each layer?

Reply:
The DOMAIN_SEPARATOR should be different anyways from the mainnet one by the call to `getChainID()` https://github.com/graphprotocol/contracts/blob/dev/contracts/disputes/DisputeManager.sol#L184

Reply:
Do you think we should keep the same salt then? To me it seemed intuitively a bit cleaner to use a new random one, but I agree the chain ID should be sufficient to get different separators so it's probably irrelevant.